[release/5.0] Fix some post-build signing issues

mmitche opened the pull request:
This fixes some post-build signing issues that are present in the aspnetcore repo
1. Add the .msi extension to be signed by Microsoft400 - Msis must be signed. With in-build signing these get handled explicitly by the wixproj infrastructure. When we do post build signing, we must sign these files.
2. Remove the strong name exclusions. These exclusions are incorrect when applied in post-build and unnecessary for in-build signing. Most importantly, the aspnetcore PKT would not end up re-strong named (it doesn't need to be strong name signed by ESRP since it's strong named in-build) because the PKT doesn't match any of the StrongNameSignInfo specified in arcade. The rest of the entries seem to be mostly about optimization. I could not find any performance difference between these entries being present and not. I am not sure whether they actually even apply to any assets. Moreover, when doing post-build signing, they would conflict with the entries in runtime and other places.

Verification - I have a tool that I wrote which unpacks every file between two directories and compares the strong name, nuget, and authenticode certs between equivalent files. This is the same tool being used to verify post-build signing. This tool shows no difference in any aspnetcore produced asset.

Baseline: https://dev.azure.com/dnceng/internal/_build/results?buildId=836183&view=results
Diff: https://dev.azure.com/dnceng/internal/_build/results?buildId=837176&view=results

Addresses #bugnumber (in this specific format)

mmitche committed:
Fix some post-build signing issues
This fixes some post-build signing issues that are present in the aspnetcore repo
1. Add the .msi extension to be signed by Microsoft400 - Msis must be signed. With in-build signing these get handled explicitly by the wixproj infrastructure. When we do post build signing, we must sign these files.
2. Remove the strong name exclusions. These exclusions are incorrect when applied in post-build and unnecessary for in-build signing. Most importantly, the aspnetcore PKT would not end up re-strong named (it doesn't need to be strong name signed by ESRP since it's strong named in-build) because the PKT doesn't match any of the StrongNameSignInfo specified in arcade. The rest of the entries seem to be mostly about optimization. I could not find any performance difference between these entries being present and not. I am not sure whether they actually even apply to any assets. Moreover, when doing post-build signing, they would conflict with the entries in runtime and other places.

Verification - I have a tool that I wrote which unpacks every file between two directories and compares the strong name, nuget, and authenticode certs between equivalent files. This is the same tool being used to verify post-build signing. This tool shows no difference in any aspnetcore produced asset.

Baseline: https://dev.azure.com/dnceng/internal/_build/results?buildId=836183&view=results
Diff: https://dev.azure.com/dnceng/internal/_build/results?buildId=837176&view=results

mmitche commented:
There is one difference as a result of the msi removal that I am investigating (will look at binlogs in PRs). The stage 1 signing submitted 10 files instead of 5. It made no meaningful difference in outputs, but I want to check the binlogs to see what the diff is there.

dougbu commented on eng/Signing.props:
```diff
@@ -57,24 +56,7 @@
-      Use the PublicKeyToken of .NET assemblies to determine which authenticode cert to use.
-      'None' is required to ensure code signing does not attempt to re-sign them or submit
-      to ESPR for strong-naming signing. We don't delay sign, so we only need to authenticode sign.
-      See https://github.com/dotnet/arcade/issues/1911 for context.
```
Let's chat tomorrow. Looks like dotnet/arcade#1911 was only closed because we had this workaround. I'd like to better understand why these exclusions aren't needed anymore.

mmitche replied:
Sounds good. I did note in my description why I thought these are unneeded:

> Remove the strong name exclusions. These exclusions are incorrect when applied in post-build and unnecessary for in-build signing. Most importantly, the aspnetcore PKT would not end up re-strong named (it doesn't need to be strong name signed by ESRP since it's strong named in-build) because the PKT doesn't match any of the StrongNameSignInfo specified in arcade. The rest of the entries seem to be mostly about optimization. I could not find any performance difference between these entries being present and not. I am not sure whether they actually even apply to any assets. Moreover, when doing post-build signing, they would conflict with the entries in runtime and other places.

mmitche committed:
Merge remote-tracking branch 'upstream/release/5.0' into remove-sn-exclusions

mmitche committed:
Do not push VS packages for installers when PostBuildSign == true

mmitche committed:
Output wix command packages to the installers output path

mmitche committed:
Don't import microbuild signing targets from wix when PostBuildSign=true

mmitche committed:
FIxup conditional

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -314,7 +314,7 @@ stages:
 
       # A few files must also go to the VS package feed.
-      - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+      - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest'), ne(variables['PostBuildSign'], 'true')) }}:
```
What sets the `PostBuildSign` variable and the corresponding property in `msbuild`❔ Similarly what sets the `$(DisableMicroBuild)` property in `msbuild`❔

I'm betting you're using these to test things out but am curious&hellip;

mmitche replied:
Right now it's a global variable in the build pipeline. It'll eventually just make its way into the YAML as a variable.

DisableMicrobuild is not set AFAIK

mmitche committed:
Tweaks:
- Don't sign wixpacks when not in post-build signing
- Generate a wixpack for both the original msi name (which the wixproj generates) AND the name we use in the final outputs. This is because while these files are the same, signing differentiates the certificate based on the file name, and wixpack lookup is also based on the file names. Aspnetcore and other repos have uses the final outputs (e.g. dotnet-aspnetcore-runtime-123.5..) as well as the internal names (e.g. AspNetCoreSharedFramework_x64.msi).
- Don't sign msi's when not post-build signing.

mmitche committed:
Avoid generating sha512 files for wixpack zips

mmitche committed:
Don't run xplat code sign jobs if PostBuildSign == true

mmitche committed:
Change original target names

mmitche committed:
Conditionalize codesign operations

mmitche committed:
Revert blazor.server.js

mmitche commented:
@dougbu Some final tweaks here:
- Because aspnetcore uses different names for the msis internally vs. publishing (and this is very difficult to resolve due to target import order), generate wixpacks for both the internal and published msi names. This allows the post build signing to connect the dots properly
- Change the names of the internal (not anything we publish to customers) names of of the ancm MSIs. This avoids an issue in publishing where the published msi and internal msi name only vary by a version number. Publishing strips away the version number and attempts to create an aka.ms link, which results in the same aka.ms link attempting to point to two different targets.
- Add the PostBuildSign flag to the YAML. this allows the template expressions to work properly
- Remove xplat code sign legs when running post build signing. This chops a significant amount of time off the build due to fewer machine allocations.

mmitche committed:
Add back the silly jobs

mmitche commented:
@dougbu I'm adding back the codesign jobs here. The aspnetcore xplat build scripts don't support arcade publishing, so the nuget packages get dropped. I think it's worth removing them (they cost a significant amount of time). but I can't in this PR.

dougbu commented:
> I think it's worth removing them

If there's a way to do this (separately), I'm interested. The codesign-xplat.yml jobs exist only to sign the non-Windows bits and steps in the Linux and macOS builds _should_ publish, albeit with manifests lacking the `-signed` suffix. So, perhaps, they could go away if MicroBuild now works on all platforms or after we flip to signing later.

Could you file an issue and label it w/ area-infrastructure if you have a suggested approach❔

mmitche commented:
I believe this is just a matter of passing --publish to arcade build scripts. I attempted, but this isn't supported by the aspnetcore custom build.sh

dougbu commented:
It's more than that @mmitche because the final steps in all non-test jobs have `/p:Publish=true` on the command line (through `$(_PublishArgs)`). The `Publish` property is all that `-publish` controls in eng/common/build.* scripts. Effectively, Linux and macOS builds publish twice with codesign-xplat.yml handling the second `-signed` one.

dougbu commented on eng/AfterSigning.targets:
```diff
@@ -17,6 +17,7 @@
       <InstallerFiles Include="$(InstallersOutputPath)**\*.tar.gz" />
       <InstallerFiles Include="$(InstallersOutputPath)**\*.wixlib" />
       <InstallerFiles Include="$(InstallersOutputPath)**\*.rpm" />
+      <InstallerFiles Remove="$(InstallersOutputPath)**\*.wixpack.zip" />
```
Curious:
1. What adds these in the first place❔
2. Why remove them❔
3. Should this be done only when signing w/in our pipeline❔

mmitche replied:
These aren't really anything for customers, so no need to generate sha512s. Reduces the overall number of output files signficantly. I changed the remove to an Exclude of wixpack.zips a couple lines above.

dougbu commented on eng/Signing.props:
```diff
@@ -5,7 +5,6 @@
   <ItemGroup>
     <!-- Reset Arcade's defaults. -->
     <ItemsToSign Remove="@(ItemsToSign)" />
-    <StrongNameSignInfo Remove="@(StrongNameSignInfo)" />
```
Think we did this to remove all of Arcade's defaults because they conflicted w/ what we wanted to sign. But, I'm fine leaving the defaults around if they don't cause problems (anymore).

mmitche replied:
This is basically what we were discussing last Friday. The defaults in arcade what we want in post-build and in-build.

dougbu commented on src/Installers/Windows/AspNetCoreModule-Setup/ANCMIISExpressV2/AncmIISExpressV2.wixproj:
```diff
@@ -7,7 +7,7 @@
     <ProjectGuid>17c76489-4c09-4e14-b81c-7a86cd937144</ProjectGuid>
     <OutputType>Package</OutputType>
-    <OutputName>ancm_iis_express_$(Platform)_en_v2</OutputName>
+    <OutputName>$(Name)_$(Platform)</OutputName>
```
@jkotalik please double-check the `*.wixproj` changes in this PR. @mmitche described the reasoning in https://github.com/dotnet/aspnetcore/pull/26506#issuecomment-705776687 but I'd appreciate another set of 👀

mmitche replied:
Yes please. I believe this should be fine though.

mmitche replied:
@jkotalik Any concerns here?

jkotalik replied:
I think it's okay, but @shirhatti and @joeloff to confirm. Pretty much, is it okay to change the name of the MSIs for IISExpress.

I think the next time we update VS, we will need to react to these name changes. besides that I think it's okay.

mmitche commented:
Why aren't the packages appearing in the first manifest though?

mmitche commented:
> Why aren't the packages appearing in the first manifest though?

I found it. Hang on, gonna give it one more attempt.

dougbu commented on src/Installers/Windows/Wix.props:
```diff
@@ -32,8 +32,8 @@
-    <!-- Directory for the zipped up light/lit command package -->
-    <WixCommandPackagesDir>$(ArtifactsNonShippingPackagesDir)</WixCommandPackagesDir>
+    <!-- Directory for the zipped up wixpack command package -->
+    <WixCommandPackagesDir>$(InstallersOutputPath)</WixCommandPackagesDir>
```
Guess that's more consistent 😺

mmitche committed:
Revert "Add back the silly jobs"

This reverts commit aa17f2404bf6e3c53351343b0a1ddc346de62e12.

dougbu reviewed (commented):
Looking better now.

Suggest reverting changes to src/Components/Web.JS/dist/Release/blazor.server.js That file just changes a bit when built on different platforms and, unless you change TypeScript files, doesn't need to be updated in GitHub.

mmitche committed:
Push nupkgs if postbuildsign == true

mmitche committed:
Remove->Exclude

dougbu commented on src/Installers/Windows/SharedFrameworkLib/SharedFrameworkLib.wixproj:
```diff
@@ -43,13 +43,19 @@
+    <ItemGroup>
+      <WixInstallerFilesToProcess Include="$(InstallersOutputPath)$(OutputName).wixlib" />
+      <WixInstallerFilesToProcess Include="$(InstallersOutputPath)$(PackageFileName)"
+        Condition=" '$(IsShipping)' == 'true' AND '$(SkipCopyToArtifactsDirectory)' != 'true' " />
```
nit: I believe `$(IsShipping)` is always `true` in this project i.e. that part of the `Condition` is only needed in Wix.targets.

dougbu reviewed (approved):
WFM once you've done your run of the custom verification tool.

dougbu commented:
BTW, a reminder
> Suggest reverting changes to src/Components/Web.JS/dist/Release/blazor.server.js

mmitche committed:
Add publishing flag for linux x64 and add deb sha512 generation

mmitche committed:
Do not push the x64 linux runtime archive more than once

mmitche committed:
true -> false

mmitche committed:
Merge remote-tracking branch 'upstream/release/5.0' into remove-sn-exclusions

mmitche committed:
Fixups
